File: ocaml/idiomatic/bin/harness.ml
(* Benchmark harness for the idiomatic OCaml engine. See spec/protocol.md §5.
   Pass L: timed per-op latency -> latencies.csv.
   Pass T: timed bulk throughput + GC allocation, then trade-stream hash + book digest. *)

open Lob

external now_ns : unit -> int = "ocaml_now_ns" [@@noalloc]

let[@inline] get_u32 s off =
  Char.code (String.unsafe_get s off)
  lor (Char.code (String.unsafe_get s (off + 1)) lsl 8)
  lor (Char.code (String.unsafe_get s (off + 2)) lsl 16)
  lor (Char.code (String.unsafe_get s (off + 3)) lsl 24)

let () =
  let in_path = if Array.length Sys.argv > 1 then Sys.argv.(1) else "bench/orders.bin" in
  let out_dir =
    if Array.length Sys.argv > 2 then Sys.argv.(2) else "bench/results/ocaml_idiomatic"
  in
  (try Sys.mkdir out_dir 0o755 with _ -> ());
  let data = In_channel.with_open_bin in_path In_channel.input_all in
  let n = String.length data / 24 in
  let warmup = n / 10 in

  (* ---- Pass L: timed per-op latency ---- *)
  let b = create ~trade_cap:(max 1 n) () in
  let lat = Array.make (max 1 n) 0 in
  for i = 0 to n - 1 do
    let off = i * 24 in
    let mtype = Char.code (String.unsafe_get data off) in
    let side = Char.code (String.unsafe_get data (off + 1)) in
    let id = get_u32 data (off + 4) in
    let price = get_u32 data (off + 8) in
    let qty = get_u32 data (off + 12) in
    let nprice = get_u32 data (off + 16) in
    let nqty = get_u32 data (off + 20) in
    let t0 = now_ns () in
    process b mtype side id price qty nprice nqty;
    lat.(i) <- now_ns () - t0
  done;
  let oc = open_out (Filename.concat out_dir "latencies.csv") in
  let buf = Buffer.create (1 lsl 16) in
  for i = warmup to n - 1 do
    Buffer.add_string buf (string_of_int lat.(i));
    Buffer.add_char buf '\n';
    if Buffer.length buf > (1 lsl 16) then begin
      output_string oc (Buffer.contents buf);
      Buffer.clear buf
    end
  done;
  output_string oc (Buffer.contents buf);
  close_out oc;

  (* ---- Pass T: timed bulk throughput + allocations ---- *)
  let b = create ~trade_cap:(max 1 n) () in
  let g0 = Gc.minor_words () in
  let mc0 = (Gc.quick_stat ()).Gc.minor_collections in
  let t0 = now_ns () in
  for i = 0 to n - 1 do
    let off = i * 24 in
    let mtype = Char.code (String.unsafe_get data off) in
    let side = Char.code (String.unsafe_get data (off + 1)) in
    let id = get_u32 data (off + 4) in
    let price = get_u32 data (off + 8) in
    let qty = get_u32 data (off + 12) in
    let nprice = get_u32 data (off + 16) in
    let nqty = get_u32 data (off + 20) in
    process b mtype side id price qty nprice nqty
  done;
  let elapsed = now_ns () - t0 in
  let minor_words = Gc.minor_words () -. g0 in
  let minor_colls = (Gc.quick_stat ()).Gc.minor_collections - mc0 in

  (* trade-stream hash + book digest (Int64, matching Rust's wrapping u64) *)
  let prime = 0x100000001b3L in
  let thash = ref 0xcbf29ce484222325L in
  let feed32 v =
    for k = 0 to 3 do
      let by = (v lsr (k * 8)) land 0xff in
      thash := Int64.mul (Int64.logxor !thash (Int64.of_int by)) prime
    done
  in
  let feedb by = thash := Int64.mul (Int64.logxor !thash (Int64.of_int by)) prime in
  for i = 0 to b.tr_n - 1 do
    feed32 b.tr_maker.(i);
    feed32 b.tr_taker.(i);
    feed32 b.tr_price.(i);
    feed32 b.tr_qty.(i);
    feedb b.tr_aggr.(i)
  done;
  let dg = digest b in
  let oc = open_out (Filename.concat out_dir "digest.txt") in
  Printf.fprintf oc "trades=%d\ntrades_hash=%016Lx\nbook_digest=%016Lx\n" b.tr_n !thash dg;
  close_out oc;

  (* percentiles for immediate feedback (analyze.py is authoritative) *)
  let m = max 0 (n - warmup) in
  let s = Array.sub lat warmup m in
  Array.sort compare s;
  let len = Array.length s in
  let pct p = if len = 0 then 0 else s.(int_of_float (float (len - 1) *. p)) in
  Printf.eprintf "messages         %d\n" n;
  Printf.eprintf "trades           %d\n" b.tr_n;
  Printf.eprintf "trades_hash      %016Lx\n" !thash;
  Printf.eprintf "book digest      %016Lx\n" dg;
  Printf.eprintf "resting orders   %d\n" (resting_count b);
  Printf.eprintf "THROUGHPUT msgs=%d elapsed_ns=%d  (%.2f M msg/s)\n" n elapsed
    (float n /. float elapsed *. 1000.);
  Printf.eprintf "ALLOC minor_words=%.0f minor_colls=%d  (%.3f words/op)\n" minor_words minor_colls
    (minor_words /. float (max 1 n));
  Printf.eprintf "latency ns  p50=%d p99=%d p999=%d max=%d\n" (pct 0.50) (pct 0.99) (pct 0.999)
    (if len > 0 then s.(len - 1) else 0)

File: rust/src/bin/harness.rs
//! Benchmark harness for the Rust engine. See `spec/protocol.md` §5.
//!
//! Three passes over the same input (deterministic, so results are identical):
//!   A. untimed, CSV sink   -> trades.csv + digest.txt   (correctness / golden artifacts)
//!   B. timed per-op        -> latencies.csv             (latency distribution)
//!   C. bulk timed          -> THROUGHPUT + ALLOC        (clean throughput + allocations)
//!
//! Usage: harness [orders.bin] [out_dir]

use lob::{Book, CountSink, Msg, Trade, TradeSink, SIDE_BID};
use std::alloc::{GlobalAlloc, Layout, System};
use std::fs::{self, File};
use std::io::{BufWriter, Read, Write};
use std::sync::atomic::{AtomicU64, Ordering::Relaxed};
use std::time::Instant;

static ALLOC_BYTES: AtomicU64 = AtomicU64::new(0);
static ALLOC_COUNT: AtomicU64 = AtomicU64::new(0);

struct Counting;
unsafe impl GlobalAlloc for Counting {
    unsafe fn alloc(&self, l: Layout) -> *mut u8 {
        ALLOC_COUNT.fetch_add(1, Relaxed);
        ALLOC_BYTES.fetch_add(l.size() as u64, Relaxed);
        System.alloc(l)
    }
    unsafe fn dealloc(&self, p: *mut u8, l: Layout) {
        System.dealloc(p, l)
    }
}
#[global_allocator]
static GLOBAL: Counting = Counting;

/// Streams trades to a CSV file and folds them into an order-sensitive FNV-1a hash
/// (the trade-stream digest used for cross-engine differential testing).
struct CsvSink<W: Write> {
    w: W,
    seq: u64,
    hash: u64,
}
impl<W: Write> TradeSink for CsvSink<W> {
    fn trade(&mut self, t: Trade) {
        const PRIME: u64 = 0x100000001b3;
        for v in [t.maker, t.taker, t.price, t.qty] {
            for b in v.to_le_bytes() {
                self.hash ^= b as u64;
                self.hash = self.hash.wrapping_mul(PRIME);
            }
        }
        self.hash ^= t.aggressor as u64;
        self.hash = self.hash.wrapping_mul(PRIME);
        let ag = if t.aggressor == SIDE_BID { 'B' } else { 'S' };
        writeln!(self.w, "{},{},{},{},{},{}", self.seq, t.maker, t.taker, t.price, t.qty, ag)
            .expect("write trade");
        self.seq += 1;
    }
}

fn main() -> std::io::Result<()> {
    let mut args = std::env::args().skip(1);
    let in_path = args.next().unwrap_or_else(|| "bench/orders.bin".to_string());
    let out_dir = args.next().unwrap_or_else(|| "bench/results/rust".to_string());
    fs::create_dir_all(&out_dir)?;

    let mut data = Vec::new();
    File::open(&in_path)?.read_to_end(&mut data)?;
    let n = data.len() / 24;
    let decode = |i: usize| Msg::decode(&data[i * 24..]);

    // ---- Pass A: untimed, produce trades.csv + digest.txt ----
    let mut book = Book::new();
    let mut sink = CsvSink {
        w: BufWriter::new(File::create(format!("{out_dir}/trades.csv"))?),
        seq: 0,
        hash: 0xcbf29ce484222325,
    };
    writeln!(sink.w, "seq,maker_id,taker_id,price,qty,aggressor")?;
    for i in 0..n {
        let m = decode(i);
        book.process(&m, &mut sink);
    }
    sink.w.flush()?;
    let digest = book.digest();
    let trades_hash = sink.hash;
    fs::write(
        format!("{out_dir}/digest.txt"),
        format!("trades={}\ntrades_hash={:016x}\nbook_digest={:016x}\n", sink.seq, trades_hash, digest),
    )?;
    let trade_count = sink.seq;

    // ---- Pass B: timed per-op latency ----
    let mut book = Book::new();
    let mut count = CountSink::default();
    let mut lat: Vec<u64> = Vec::with_capacity(n);
    for i in 0..n {
        let m = decode(i);
        let t0 = Instant::now();
        book.process(&m, &mut count);
        lat.push(t0.elapsed().as_nanos() as u64);
    }
    let warmup = n / 10;
    let mut lw = BufWriter::new(File::create(format!("{out_dir}/latencies.csv"))?);
    for &v in &lat[warmup..] {
        writeln!(lw, "{v}")?;
    }
    lw.flush()?;

    // ---- Pass C: clean throughput + allocations (no per-op clock) ----
    let mut book = Book::new();
    let mut count = CountSink::default();
    let ab0 = ALLOC_BYTES.load(Relaxed);
    let ac0 = ALLOC_COUNT.load(Relaxed);
    let start = Instant::now();
    for i in 0..n {
        let m = decode(i);
        book.process(&m, &mut count);
    }
    let elapsed_ns = start.elapsed().as_nanos() as u64;
    let alloc_bytes = ALLOC_BYTES.load(Relaxed) - ab0;
    let alloc_count = ALLOC_COUNT.load(Relaxed) - ac0;

    // quick percentiles for immediate feedback (analyze.py is authoritative)
    let mut s = lat[warmup..].to_vec();
    s.sort_unstable();
    let pct = |p: f64| s[((s.len() as f64 - 1.0) * p) as usize];

    eprintln!("messages         {n}");
    eprintln!("trades           {trade_count}");
    eprintln!("trades_hash      {trades_hash:016x}");
    eprintln!("book digest      {digest:016x}");
    eprintln!("resting orders   {}", book.resting_count());
    eprintln!("THROUGHPUT msgs={n} elapsed_ns={elapsed_ns}  ({:.2} M msg/s)", n as f64 / elapsed_ns as f64 * 1000.0);
    eprintln!("ALLOC bytes={alloc_bytes} allocs={alloc_count}  ({:.3} bytes/op)", alloc_bytes as f64 / n as f64);
    eprintln!(
        "latency ns  p50={} p99={} p999={} max={}",
        pct(0.50),
        pct(0.99),
        pct(0.999),
        s[s.len() - 1]
    );
    Ok(())
}
